### PR TITLE
home-work

### PR DESCRIPTION
**Pull request by ddryzhov**
Added spring-boot-starter-security, SecurityConfig class with PasswordEncoder and SecurityFilterChain, Implemented UserDetails interface in the User class, Added the implementation for UserDetailsService, Added Role entity, RoleRepository, Liquibase changesets, Used @PreAuthorize

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -93,6 +93,11 @@
             <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
             <version>2.3.0</version>
         </dependency>
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+        </dependency>
```
Comment (Vasul-Pankiv) [inline pom.xml:88]: remove the empty line before dependency or  add after

File: src/main/java/mate/academy/bookstore/controller/BookController.java [hunk 2]
```diff
@@ -59,26 +62,30 @@ public BookDto getBookById(@PathVariable @Positive Long id) {
         return bookService.findById(id);
     }
 
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstore/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasRole('ADMIN')")
```
Spring will add this prefix by itself

File: src/main/java/mate/academy/bookstore/controller/BookController.java [hunk 3]
```diff
@@ -37,6 +38,7 @@
 public class BookController {
     private final BookService bookService;
 
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstore/controller/BookController.java]: check for USER role is redundant

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 4]
```diff
@@ -35,4 +47,52 @@ public class User {
...
+    }
+
+    @Override
+    public String getUsername() {
+        return email;
+    }
+
+    @Override
+    @NotBlank
+    @Length(min = 8)
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 5]
```diff
@@ -2,4 +2,16 @@ databaseChangeLog:
...
+  - include:
+      file: db/changelog/changes/04-create-roles-table.yaml
+  - include:
+      file: db/changelog/changes/05-create-users-roles-table.yaml
+  - include:
+      file: db/changelog/changes/06-insert-roles.yaml
+  - include:
+      file: db/changelog/changes/07-assign-roles.yaml
+  - include:
+      file: db/changelog/changes/08-add-is-deleted-column-to-users.yaml
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: file should end with empty line

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 6]
```diff
@@ -35,4 +47,50 @@ public class User {
     private String lastName;
 
     private String shippingAddress;
+
+    @Column(name = "is_deleted")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 7]
```diff
@@ -35,4 +47,50 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/mate/academy/bookstore/repository/role/RoleRepository.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.bookstore.repository.role;
+
+import mate.academy.bookstore.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/role/RoleRepository.java:8]:
```suggestion
@Repository
public interface RoleRepository extends JpaRepository<Role, Long> {
```

File: src/main/java/mate/academy/bookstore/repository/user/UserRepository.java [hunk 9]
```diff
@@ -1,10 +1,15 @@
...
 import mate.academy.bookstore.model.User;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
+
+    @Query("SELECT u FROM User u LEFT JOIN FETCH u.roles WHERE u.email = :email")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/user/UserRepository.java]:
```suggestion
    @EntityGraph(attributePaths = "roles")
```

File: src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() ->
+                        new UsernameNotFoundException("User not found with email: " + email));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/CustomUserDetailsService.java]:
```suggestion
                        new EntityNotFoundException("User not found with email: " + email));
```

File: src/main/resources/db/changelog/changes/03-insert-users.yaml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+                  value: "${hashedUserPassword}"
+              - column:
+                  name: first_name
+                  value: "Regular"
+              - column:
+                  name: last_name
+                  value: "User"
+              - column:
+                  name: shipping_address
+                  value: "123 User St"
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/03-insert-users.yaml:42]:
```suggestion
                  value: "123 User St"
                  
```

File: src/main/resources/db/changelog/changes/04-create-roles-table.yaml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: role
+                  type: ENUM('ROLE_USER', 'ROLE_ADMIN')
+                  constraints:
+                    nullable: false
+                    unique: true
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-roles-table.yaml:21]:
```suggestion
                    unique: true
                    
```

File: src/main/resources/db/changelog/changes/05-create-users-roles-table.yaml [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+                  name: user_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+        - addPrimaryKey:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/05-create-users-roles-table.yaml]:
```suggestion
```

File: src/main/resources/db/changelog/changes/05-create-users-roles-table.yaml [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+            baseColumnNames: user_id
+            constraintName: fk_users_roles_user
+            referencedTableName: users
+            referencedColumnNames: id
+        - addForeignKeyConstraint:
+            baseTableName: users_roles
+            baseColumnNames: role_id
+            constraintName: fk_users_roles_role
+            referencedTableName: roles
+            referencedColumnNames: id
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/05-create-users-roles-table.yaml:30]:
```suggestion
            referencedColumnNames: id
            
```

File: src/main/resources/db/changelog/db.changelog-master.yaml (hunk 5 above)
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/db.changelog-master.yaml:24]: add empty line at the end of every changelogs

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

**Comment by ddryzhov** [comment]
I didn't add an empty line after the liquibase files because I already had it added, I don't know why it doesn't show up in git, but I can add two empty lines :-) May I ask why we need to use RoleRepository if we don't use it anywhere? Is this for the future or do I have a mistake?

File: src/main/resources/db/changelog/db.changelog-master.yaml (hunk 5 above)
Comment (ddryzhov) [reply src/main/resources/db/changelog/db.changelog-master.yaml:24]: May I ask why we need to use RoleRepository if we don't use it anywhere? Is this for the future or do I have a mistake?

**Review by sarakhmen — APPROVED** [review]
Good job! You'll need rolerepository later :)

**Pull request merged**